### PR TITLE
Add optimization result collection to basic result collection

### PR DESCRIPTION
**Pull request by SimonBoothroyd**
## Description
This PR adds a convenience method to create a basic data set collection from an existing optimization result collection. This would allow, for example, being able to find all of the hessian result records which were generated using the final structure of a set of optimizations.

This may be preferable in certain cases to just creating the basic result collection itself. This is especially true when trying to create a result collection from an older dataset which does not have CMILES information attached. This method will propagate the parent optimization CMILES down to the child basic result record thus circumventing this issue.

## Status
- [x] Ready to go

### CODE REVIEW

**Review by jthorton — COMMENTED** [review]
Thanks for taking this on, I think this is quite clean and very useful. My one concern is that I don't think we are checking the qcspec/keywords match those used for the optimisation dataset. You can include this in the query see [here](https://github.com/MolSSI/QCFractal/blob/5f00dd06bb34ca912c4055f0cbac60863ea89c7f/qcfractal/interface/client.py#L651) which saves us doing the filtering. I think its safe to assume users have used the same program, method, basis and keywords for both calculations so we should filter by these as well.

File: openff/qcsubmit/results/caching.py [hunk 1]
```diff
@@ -102,11 +102,11 @@ def _cached_client_query(
         if (client_address, query_id) in query_cache
     ]
 
-    client = _cached_fractal_client(client_address)
+    client = cached_fractal_client(client_address)
 
     logger.debug(f"starting {query_name} to {client_address}")
 
-    batch_query_ids = _batched_indices(missing_query_ids, client.query_limit)
+    batch_query_ids = batched_indices(missing_query_ids, client.query_limit)
```
Comment (jthorton) [inline openff/qcsubmit/results/caching.py:109]: Just wondering what the names have changed?

Comment (SimonBoothroyd) [reply openff/qcsubmit/results/caching.py:109]: I'm starting to access these a bit outside of the `caching` module so I thought it was probably ok to just make them public members. I'd be happy to make them private though if that's your preference?

**Comment by SimonBoothroyd** [comment]
> My one concern is that I don't think we are checking the qcspec/keywords match those used for the optimisation dataset. You can include this in the query see here which saves us doing the filtering. I think its safe to assume users have used the same program, method, basis and keywords for both calculations so we should filter by these as well.

Ah brilliant catch - thanks! I've (I think) added this in 02db5a6

**Review by jthorton — APPROVED** [review]
Thanks for adding the specification checks. LGTM!

**Pull request merged**